[Jetpack Performance] Add local feature flag and enable app passwords for all endpoints

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-1220

### Description
This PR adds the following changes:
- Refactors the handling of app passwords configuration, so instead of just supplying the `ApplicationClientId` to the DI, we now supply an implementation of `ApplicationPasswordsConfiguration`, this allows us to control whether the feature is enabled or not from the app module.
- Add a feature flag and use it to enable the feature. We'll replace this with a remote feature flag.
- Move the `WooExperimentalNetwork` logic to `WooNetwork`, and update it to use the new property of `ApplicationPasswordsConfiguration`.
- Improve the logging of failures (wip for now, I'll revisit this to see if we need further improvements).

### Steps to reproduce
1. Confirm the feature flag `APP_PASSWORDS_FOR_JETPACK_SITES` is enabled.
2. Use a way to inspect app network requests.
4. Open the app.
5. Make sure you are signed in using WordPress.com account.
6. Confirm the app is using app passwords (talking to the site directly using its URL, and authenticating using app passwords.
7. Disable the feature flag `APP_PASSWORDS_FOR_JETPACK_SITES`.
8. Re-open the app.
9. Confirm the app is using Jetpack proxy.

### Testing information
The above steps should be enough.

### The tests that have been performed
The above.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Refactor handling of Application Passwords configuration

This gives the clients more freedom in enabling disabling features, which will allow us to control the feature using remote feature flag.

hichamboushaba committed:
Rename property

hichamboushaba committed:
Add new property to control whether feature is enabled or not for WPCom login

hichamboushaba committed:
Copy `WooExperimentalNetwork` logic to `WooNetwork`

hichamboushaba committed:
Force using Jetpack tunnel when feature is disabled

hichamboushaba committed:
Remove WooExperimentalNetwork

hichamboushaba committed:
Remove unused argument

hichamboushaba committed:
Remove unused remote config key

hichamboushaba committed:
Remove old code related to the initial experiment

hichamboushaba committed:
Improve log message for failed requests

hichamboushaba committed:
Add a feature flag for the feature

dangermattic commented:
<!--
  0 Errors
  3 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>WooWearApplicationPasswordsConfiguration</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>WooApplicationPasswordsConfiguration</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Minor refactoring for the configuration class

- Renames the functions for better clarity.
- Make `isEnabledForJetpackAccess` suspendable to be ready for remote feature flag access

hichamboushaba committed:
Update unit tests

hichamboushaba committed:
Make `isEnabledForDirectAccess` enabled by default

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix wear app build issue

hichamboushaba commented on WooCommerce-Wear/src/main/java/com/woocommerce/android/wear/di/ApplicationPasswordsModule.kt:
```diff
@@ -0,0 +1,24 @@
+class WooWearApplicationPasswordsConfiguration @Inject constructor() : ApplicationPasswordsConfiguration {
+    override val applicationName: String = ""
+
+    override fun isEnabledForDirectAccess(): Boolean = false
+    override suspend fun isEnabledForJetpackAccess(): Boolean = false
```
The wear app didn't support Application Passwords before https://github.com/woocommerce/woocommerce-android/pull/12124, and we are keeping the same behavior.

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpapi/applicationpasswords/ApplicationPasswordsConfiguration.kt:
```diff
@@ -1,26 +1,17 @@
+     * Returns true if the Application Passwords feature is enabled for Jetpack-mediated access.
+     * This applies when sites are accessed through Jetpack, even if they are self-hosted.
+     */
+    suspend fun isEnabledForJetpackAccess(): Boolean
```
Having this a suspendable function will make using remote feature flags easier for us, given they require DB access.

hichamboushaba committed:
Fix issue for Jetpack CP sites

`isUsingWpComRestApi` was returning false for Jetpack CP sites, leading to always ignoring saved app passwords credentials, as for WPCom rest sites, `SiteModel#username` is always null.

hichamboushaba commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/model/SiteModel.java:
```diff
@@ -787,7 +787,7 @@ public void setOrigin(@SiteOrigin int origin) {
 
     public boolean isUsingWpComRestApi() {
-        return isWPCom() || (isJetpackConnected() && getOrigin() == ORIGIN_WPCOM_REST);
+        return isWPCom() || getOrigin() == ORIGIN_WPCOM_REST;
```
When testing the app passwords feature with a Jetpack CP site, I found an issue, we always ignore the created password, and the cause was in this condition.
I checked the usages of this function, and I think we can safely remove the Jetpack connection condition here, as they apply also to Jetpack CP sites.

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/WooNetwork.kt:
```diff
@@ -1,25 +1,36 @@
  * The [SiteModel.ORIGIN_XMLRPC] support is kept for backward compatibility
  */
 @Singleton
 class WooNetwork @Inject constructor(
```
All the changes here are a copy of what we had before in `WooExperimentalNetwork`

hichamboushaba committed:
Add preference key for the experimental toggle

hichamboushaba committed:
Add a preference toggle to the experimental features screen

hichamboushaba committed:
Rename settings entry to "Experimental features"

To match iOS platform

hichamboushaba commented:
@irfano since Jorge is AFK for two days, I'm removing his assignment for this PR.
Reviewing this is not urgent, but if you have some time you'll have something to review, and this is a base for my other PRs, so having it merged before them will make things easier for me.

irfano commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/WooNetwork.kt:
```diff
@@ -178,4 +175,21 @@ class WooNetwork @Inject constructor(
+                "path: ${requestContext.path},\n" +
+                "method: ${requestContext.method},\n" +
+                "error: HTTP status code ${error.volleyError.networkResponse?.statusCode}, " +
+                "error message: ${error.errorCode?.ifEmpty { null } ?: error.message}",
```
Nice logging details, they’ll be helpful for debugging. 👍🏻

irfano reviewed (approved):
It works expected. I also tested the Wear app, it works there as well. I'm approving but not merging in case you'd like to make changes based on these notes:
- In [WPAPIResponse](https://github.com/woocommerce/woocommerce-android/blob/0b27a820c9cf89574079a4afc91129182c2e5eaa/libs/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpapi/WPAPIResponse.kt#L21), there's this comment: _"This is used for tracking purposes, and used only with the Woo experimental networking mode."_ I believe this is no longer valid and can be removed.
- I believe the [fetchAppPasswordsSupportIfNeeded()](https://github.com/woocommerce/woocommerce-android/blob/6b6148265931b302fb2e91b664b5ec5530fad0c8/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WooCommerceStore.kt#L114) function is no longer needed and can be removed.

irfano committed:
Merge pull request #14559 from woocommerce/issue/WOOMOB-1183-jp-app-passwords-toggle

[Jetpack Performance] Add a preference toggle for enabling app passwords for Jetpack sites

hichamboushaba committed:
Remove outdated comment section

hichamboushaba committed:
Merge branch 'trunk' into issue/WOOMOB-1126-app-passwords-network

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt

hichamboushaba committed:
Remove old way for checking for app passwords support

hichamboushaba commented:
Thanks @irfano, both remarks are addressed, nice catch.
